File: apps/server/src/research/stats/dtos.rs
use crate::academic::AcademicOption;
use crate::research::JournalKind;

use serde::{Deserialize, Serialize};
use uuid::Uuid;
use validator::Validate;

#[derive(Debug, Deserialize, Validate, Default)]
#[serde(rename_all = "camelCase")]
pub struct WorksStatsQuery {
	pub journal_kind: Option<JournalKind>,

	pub department_id: Option<Uuid>,

	#[validate(range(min = 1900, max = 2100))]
	pub year_from: Option<i16>,

	#[validate(range(min = 1900, max = 2100))]
	pub year_to: Option<i16>,

	#[validate(range(min = 1, max = 100))]
	pub limit: Option<i32>,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct YearValue {
	pub year: i16,
	pub value: i64,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct TimeSeriesStat {
	pub id: Option<String>,
	pub key: String,
	pub values: Vec<YearValue>,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ScopeSeries {
	pub id: Option<String>,
	pub name: String,
	pub total: i64,
	pub wos: Vec<YearValue>,
	pub scopus: Vec<YearValue>,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ScopeTotal {
	pub id: Option<String>,
	pub name: String,
	pub total: i64,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct FacultySummary {
	pub total_works: i64,
	pub wos_count: i64,
	pub scopus_count: i64,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct WorksStatsResponse {
	pub faculty_summary: FacultySummary,
	pub by_journal_kind: Vec<TimeSeriesStat>,
	pub by_department: Vec<ScopeSeries>,
	pub by_research_line: Vec<ScopeSeries>,
	pub top_publishers: Vec<TopPublisher>,
}

#[derive(Debug, Deserialize, Validate, Default)]
#[serde(rename_all = "camelCase")]
pub struct DepartmentDetailQuery {
	#[validate(range(min = 1900, max = 2100))]
	pub year_from: Option<i16>,

	#[validate(range(min = 1900, max = 2100))]
	pub year_to: Option<i16>,

	pub option: Option<AcademicOption>,

	pub journal_kind: Option<JournalKind>,

	#[validate(range(min = 1, max = 100))]
	pub limit: Option<i32>,
}

#[derive(Debug, Deserialize, Validate, Default)]
#[serde(rename_all = "camelCase")]
pub struct AcademicStatsQuery {
	#[validate(range(min = 1900, max = 2100))]
	pub year_from: Option<i16>,

	#[validate(range(min = 1900, max = 2100))]
	pub year_to: Option<i16>,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ResearchLineStat {
	pub research_line_id: String,
	pub name: String,
	pub count: i64,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct AcademicContribution {
	pub academic_works: i64,
	pub faculty_works: i64,
	pub department_works: i64,
	pub department_name: Option<String>,
	pub dominant_line_works: i64,
	pub line_total_works: i64,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct AcademicStatsResponse {
	pub by_research_line: Vec<ResearchLineStat>,
	pub dominant_research_line_id: Option<String>,
	pub by_journal_kind: Vec<TimeSeriesStat>,
	pub contribution: AcademicContribution,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct TopPublisher {
	pub academic_id: String,
	pub name: String,
	pub total: i64,
	pub scopus: i64,
	pub wos: i64,
	pub unindexed: i64,
	pub option: String,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct DepartmentDetailResponse {
	pub department: String,
	pub total_works: i64,
	pub scopus_count: i64,
	pub wos_count: i64,
	pub teaching_count: i64,
	pub research_count: i64,
	pub by_journal_kind: Vec<TimeSeriesStat>,
	pub top_publishers: Vec<TopPublisher>,
}

#[derive(Debug, Deserialize, Validate, Default)]
#[serde(rename_all = "camelCase")]
pub struct ResearchLineStatsQuery {
	#[validate(range(min = 1900, max = 2100))]
	pub year_from: Option<i16>,

	#[validate(range(min = 1900, max = 2100))]
	pub year_to: Option<i16>,

	#[validate(range(min = 1, max = 100))]
	pub limit: Option<i32>,
}

#[derive(Debug, Clone, Copy, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub enum ProductivityDegree {
	All,
	Magister,
	Doctor,
}

#[derive(Debug, Clone, Copy, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub enum ProductivityScope {
	Faculty,
	Department,
	ResearchLine,
}

#[derive(Debug, Clone, Copy, Deserialize, Serialize, Default)]
#[serde(rename_all = "camelCase")]
pub enum ProductivityJceScope {
	#[default]
	Doctor,
	All,
}

#[derive(Debug, Deserialize, Validate, Default)]
#[serde(rename_all = "camelCase")]
pub struct ProductivityQuery {
	pub degree: Option<ProductivityDegree>,
	pub scope: Option<ProductivityScope>,
	pub jce_scope: Option<ProductivityJceScope>,
	pub department_id: Option<Uuid>,
	pub research_line_id: Option<Uuid>,

	#[validate(range(min = 1, max = 12))]
	pub month: Option<i16>,

	#[validate(range(min = 1900, max = 2100))]
	pub year_from: Option<i16>,

	#[validate(range(min = 1900, max = 2100))]
	pub year_to: Option<i16>,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ProductivityYearValue {
	pub year: i16,
	pub value: f64,
	pub pubs: i64,
	pub jce: f64,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ProductivitySeries {
	pub key: String,
	pub values: Vec<ProductivityYearValue>,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ProductivityResponse {
	pub trend: Vec<ProductivitySeries>,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ResearchLineStatsResponse {
	pub name: String,
	pub total_works: i64,
	pub wos_count: i64,
	pub scopus_count: i64,
	pub by_journal_kind: Vec<TimeSeriesStat>,
	pub by_department: Vec<ScopeTotal>,
	pub top_publishers: Vec<TopPublisher>,
}

File: apps/client/src/lib/research/stats/dtos.ts
import * as v from "valibot"

export interface StatsQuery {
	journalKind?: "wos" | "scopus"
	departmentId?: string
	yearFrom?: number
	yearTo?: number
	limit?: number
}

export interface YearValue {
	year: number
	value: number
}

export interface TimeSeriesStat {
	id?: string | null
	key: string
	values: YearValue[]
}

export interface ScopeSeries {
	id: string | null
	name: string
	total: number
	wos: YearValue[]
	scopus: YearValue[]
}

export interface ScopeTotal {
	id: string | null
	name: string
	total: number
}

export interface FacultySummary {
	totalWorks: number
	wosCount: number
	scopusCount: number
}

export interface WorksStatsResponse {
	facultySummary: FacultySummary
	byJournalKind: TimeSeriesStat[]
	byDepartment: ScopeSeries[]
	byResearchLine: ScopeSeries[]
	topPublishers: TopPublisher[]
}

export interface TopPublisher {
	academicId: string
	name: string
	total: number
	scopus: number
	wos: number
	unindexed: number
	option: string
}

export interface DepartmentDetail {
	department: string
	totalWorks: number
	scopusCount: number
	wosCount: number
	teachingCount: number
	researchCount: number
	byJournalKind: TimeSeriesStat[]
	topPublishers: TopPublisher[]
}

export interface DepartmentDetailQuery {
	yearFrom?: number
	yearTo?: number
	option?: "teaching" | "research"
	journalKind?: "wos" | "scopus"
	limit?: number
}

export interface AcademicStatsQuery {
	yearFrom?: number
	yearTo?: number
}

export interface ResearchLineStat {
	researchLineId: string
	name: string
	count: number
}

export interface AcademicContribution {
	academicWorks: number
	facultyWorks: number
	departmentWorks: number
	departmentName: string | null
	dominantLineWorks: number
	lineTotalWorks: number
}

export interface AcademicStatsResponse {
	byResearchLine: ResearchLineStat[]
	dominantResearchLineId: string | null
	byJournalKind: TimeSeriesStat[]
	contribution: AcademicContribution
}

export interface ResearchLineStatsQuery {
	yearFrom?: number
	yearTo?: number
	limit?: number
}

export interface ResearchLineStatsResponse {
	name: string
	totalWorks: number
	wosCount: number
	scopusCount: number
	byJournalKind: TimeSeriesStat[]
	byDepartment: ScopeTotal[]
	topPublishers: TopPublisher[]
}

export type ProductivityDegree = "all" | "magister" | "doctor"
export type ProductivityScope = "faculty" | "department" | "researchLine"
export type ProductivityJceScope = "doctor" | "all"

export interface ProductivityQuery {
	degree?: ProductivityDegree
	scope?: ProductivityScope
	jceScope?: ProductivityJceScope
	departmentId?: string
	researchLineId?: string
	month?: number
	yearFrom?: number
	yearTo?: number
}

export interface ProductivityYearValue {
	year: number
	value: number
	pubs: number
	jce: number
}

export interface ProductivitySeries {
	key: string
	values: ProductivityYearValue[]
}

export interface ProductivityResponse {
	trend: ProductivitySeries[]
}

// Stats search params schema

const currentYear = new Date().getFullYear()
const defaultYearFrom = String(currentYear - 5)
const defaultYearTo = String(currentYear)

export const statsParamsDTOSchema = v.object({
	yearFrom: v.optional(v.fallback(v.string(), defaultYearFrom), defaultYearFrom),
	yearTo: v.optional(v.fallback(v.string(), defaultYearTo), defaultYearTo),
	limit: v.optional(v.fallback(v.string(), "10"), "10"),
})
